src/AspNet.Security.OAuth.Shopify/ShopifyAuthenticationHandler.cs

```diff
@@ -0,0 +1,218 @@
+                                /* Status: */ response.StatusCode,
+                                /* Headers: */ response.Headers.ToString(),
+                                /* Body: */ await response.Content.ReadAsStringAsync());
+            
+                throw new HttpRequestException("An error occurred while retrieving the shop profile.");
+            }
+
+            var payload = JObject.Parse(await response.Content.ReadAsStringAsync());
+
+#if DEBUG
+            // ReSharper disable once UnusedVariable
+            var jsonStr = payload.ToString(Formatting.Indented);
+#endif
+
+            // In Shopify, the customer can modify the scope given to the app. Apps should verify
+            // that the customer is allowing the required scope.
+            var actualScope = tokens.Response["scope"].ToString();
+            var isPersistent = true;
+
+            // If the request was for a "per-user" (i.e. no offline access)
+            if (tokens.Response.TryGetValue("expires_in", out var val))
+            {
+                isPersistent = false;
+
+                var expires = DateTimeOffset.Now.AddSeconds(Convert.ToInt32(val.ToString()));
+                identity.AddClaim(new Claim(ClaimTypes.Expiration, expires.ToString("O"), ShopifyAuthenticationOptions.XmlSchemaDateTime));
+
+                actualScope = tokens.Response["associated_user_scope"].ToString();
+                
+                var userData = tokens.Response["associated_user"].ToString();
+                identity.AddClaim(new Claim(ClaimTypes.UserData, userData));
+            }
+
+            identity.AddClaim(new Claim(ClaimTypes.IsPersistent, isPersistent.ToString(), ShopifyAuthenticationOptions.XmlSchemaBoolean));
+            identity.AddClaim(new Claim(ShopifyAuthenticationDefaults.ShopifyScopeClaimType, actualScope));
+
+            var principal = new ClaimsPrincipal(identity);
+
+            var context = new OAuthCreatingTicketContext(principal, properties, Context, Scheme, Options, Backchannel, tokens, payload);
+            context.RunClaimActions(payload);
+
+            await Options.Events.CreatingTicket(context);
+            var ticket = new AuthenticationTicket(context.Principal, context.Properties, Scheme.Name);
+
+            return ticket;
+        }
+
+        /// <inheritdoc />
+        protected override string FormatScope()
+        {
+            return string.Join(",", Options.Scope);
+        }
+
+        /// <inheritdoc />
+        protected override string BuildChallengeUrl(AuthenticationProperties properties, string redirectUri)
+        {
+            if (!properties.Items.ContainsKey(ShopifyAuthenticationDefaults.ShopNameAuthenticationProperty))
+            {
+                var msg =
+                    $"Shopify provider AuthenticationProperties must contain {ShopifyAuthenticationDefaults.ShopNameAuthenticationProperty}";
```

Comment:
`"The Shopify provider's AuthenticationProperties must contain the {ShopifyAuthenticationDefaults.ShopNameAuthenticationProperty} property."`?